### ISSUE TITLE
Option to disable `no-non-null-assertion` for React refs

### ISSUE THREAD

**Issue by SevenOutman** [typescript-eslint/typescript-eslint#4161]
Hi, sorry for not follow a template but I didn't see a template suitable for suggesting rule options.

I've been using `@typescript-eslint` in my React TypeScript projects. I have this common scenario where I create a ref with `useRef()` hook and bind it with a DOM element.

```jsx
const divRef = useRef<HTMLDivElement>(null);

return <div ref={divRef} />
```

React makes sure that `divRef.current` won't be null after the first render, so I have to add `!` to every `divRef.current!` calls to suppress TS warning about possible null values.
It's such a daily scenario so I wonder if there is a way of disabling this rule just for React refs. Thanks.

**Comment by bradzacher** [comment]
First up - please.make sure you use the issue wizard as well as flowing the issue template. Bypassing the templates means you could be excluding information we need.

Make sure you click "new issue" and then select one of the templates.

----

We do not encode framework-specific semantics into our rules. We are a general purpose project, and supporting individual frameworks or libraries is outside our scope. Thus we would not build anything specific for "react".

Additionally - even if we would - it's difficult to properly detect this case statically. Making the cost-benefit low.

Finally, the null case is completely valid and something you should handle. If someone were to say - delete the props usage - your component would crash entirely because you've unsafely consumed the ref pointer.
Or OFC there is the case of the very first render, where it will be `null`.

Optional chaining is a great way to handle this - or practice defensive programming and guard against the null case explicitly.
